### PR TITLE
perf(ivf): keep routing statistics parsing out of ClassifyDatas reduce lock

### LINKED ISSUE #2825: [improve](ivf): keep routing statistics parsing out of the ClassifyDatas reduce lock
### Kind of improvement

perf

### Affected area

ivf

## Summary

`IVFNearestPartition::ClassifyDatas` holds `dist_cmp_reduce_mutex` across a full
`DatasetImpl::GetStatistics` call, so an allocation-heavy JSON round trip runs
inside the critical section once per routed vector.

`src/algorithm/ivf/ivf_nearest_partition.cpp:117-126` (upstream `main`):

```cpp
std::scoped_lock lock(dist_cmp_reduce_mutex);
// the return value of GetStatistics always has the same length as the input keys, and
// atoi("") returns a `0`.
auto route_stats = search_result->GetStatistics({"dist_cmp", "distance_evaluations"});
dist_cmp += std::atoi(route_stats[0].c_str());
if (ctx != nullptr and ctx->stats != nullptr and route_stats.size() > 1) {
    ctx->stats->AddDistance(SearchStatistics::DistancePhase::ROUTING,
                            DistanceEvaluationBackend::FP32,
                            std::strtoull(route_stats[1].c_str(), nullptr, 10));
}
```

`GetStatistics` re-parses the statistics JSON the routing HGraph had just
serialized, then dumps two values back into strings for the caller to `atoi`
(`src/dataset_impl.cpp:602-612`):

```cpp
DatasetImpl::GetStatistics(const std::vector<std::string>& stat_keys) const {
    auto json = JsonType::Parse(this->Statistics_);
    ...
        result.emplace_back(json[key].Dump());
```

For the routing HGraph the payload is ~820 bytes; parsing it builds a
`nlohmann` `std::map` DOM of roughly 40 red-black-tree nodes, each with a key
`std::string`, i.e. about 40 heap allocations, and the two `Dump()` calls
allocate again. All of it is serialized by the mutex.

Two observations make this avoidable:

1. The build path calls `ClassifyDatas` with `ctx == nullptr`
   (`src/algorithm/ivf/ivf.cpp:433`), as does the bucket datacell
   (`src/datacell/bucket_datacell.h:609`), and the accumulated `dist_cmp` is only
   read when `ctx` and `ctx->stats` are both non-null
   (`src/algorithm/ivf/ivf_nearest_partition.cpp:142-144`). The whole accounting
   block is therefore dead on the build path and can return early.
2. On the search path the parsing does not need the lock at all. Only the two
   integer accumulations do.

## Expected benefit

Runtime perf on IVF build. The critical section shrinks from a ~26-37 us
allocation-heavy JSON round trip per routed vector to two integer additions,
which is what currently keeps an 8-thread build from using more than ~2.5-2.8
cores.

There is no behaviour change: the same `dist_cmp` and
`SearchStatistics::DistancePhase::ROUTING` values are produced on the search
path, and the build path never consumed them.

## Benchmark evidence

Measured with a patched vs unpatched build of the same tree, IVF `Add` phase
only, on 200k x 128d random fp32 vectors, `buckets_count=256`,
`build_thread_count=8`, 8 physical cores. "core" is the average number of cores
actually occupied, derived from summing `utime + stime` over
`/proc/self/task/*/stat` and dividing by wall time.

| allocator | before | after |
| --- | --- | --- |
| glibc malloc | 14.16 s / 2.49 core | 7.02 s / 5.12 core |
| tcmalloc | 8.56 s / 2.79 core | 2.08 s / 7.50 core |

The allocator matters because the work inside the lock is allocation-bound;
tcmalloc removes the arena contention that otherwise masks the win.

Attribution was done by sampling stacks with gdb (`thread apply all bt`) during
the build: 117 of 360 sampled frames sat on the `std::scoped_lock` at
`ivf_nearest_partition.cpp:117`.

Recall is unaffected (routing decisions are untouched); peak RSS is unaffected.

## References

- `src/algorithm/ivf/ivf_nearest_partition.cpp:96` - `dist_cmp_reduce_mutex` declaration
- `src/algorithm/ivf/ivf_nearest_partition.cpp:117-126` - the critical section
- `src/algorithm/ivf/ivf_nearest_partition.cpp:142-144` - where `dist_cmp` is consumed
- `src/algorithm/ivf/ivf.cpp:433` and `src/datacell/bucket_datacell.h:609` - build paths passing `ctx == nullptr`
- `src/dataset_impl.cpp:602-612` - `GetStatistics` parse-and-redump

## Additional context

The deeper cause is that `InnerIndexInterface::KnnSearch` has no overload taking
a `QueryContext*`, so an internal caller cannot read statistics without going
through the string channel (serialize to JSON, parse back, `atoi`). Narrowing
the lock fixes the IVF symptom; an internal overload that passes the context
directly would remove the round trip for every internal caller. Happy to file
that separately if maintainers think it is worth doing.

## Related issues

Searched `--state all` for `ClassifyDatas`, `GetStatistics lock`. No duplicate
found. Adjacent work on IVF build performance:

- #2812 `[feat](ivf): offload IVF index build to an optional CUDA backend` (open)

## Environment

```
- OS: Alibaba Group Enterprise Linux Server 7.2
- vsag version: based on upstream main @ 5096a0a6
- compiler version: c++ (GCC) 9.2.1 20240328
```

_Drafted with AI assistant: Qoder:claude-sonnet-4.6_

### PR DESCRIPTION
**Pull request by LHT129**
## Summary

`IVFNearestPartition::ClassifyDatas` held `dist_cmp_reduce_mutex` across a full `DatasetImpl::GetStatistics` call once per routed vector. `GetStatistics` re-parses the routing HGraph statistics JSON and re-dumps two values back to strings for `atoi`/`strtoull`, an allocation-heavy round trip that ran inside the critical section and throttled multi-threaded IVF builds.

This PR narrows the critical section:

- Skip the whole accounting block when `ctx`/`ctx->stats` are null — the build path (`ivf.cpp` `ClassifyDatas(..., nullptr)`, `bucket_datacell.h` `Train`) never consumes these values, so this is dead-code removal.
- On the search path, parse `GetStatistics` outside the lock; only the two integer accumulations (`dist_cmp += ...` and `AddDistance(...)`) are serialized. `SearchStatistics::AddDistance` is already lock-free (atomics + CAS), so the lock now only guards the local `dist_cmp` counter.

Behavior is unchanged: identical `dist_cmp` and `SearchStatistics::DistancePhase::ROUTING` values on the search path, and the build path never produced them.

## Changes

- `src/algorithm/ivf/ivf_nearest_partition.cpp`: move the statistics parse out of `dist_cmp_reduce_mutex`; skip it entirely when `ctx`/`stats` are null.
- `src/algorithm/ivf/ivf_nearest_partition_test.cpp`: add "IVF Nearest Partition Routing Statistics Test" covering the search path (non-null `QueryContext` with stats) for both pooled and non-pooled thread configurations, asserting identical bucket assignment and non-zero routing stats.

## Testing

- `unittests [IVFNearestPartition]` — 2010 assertions in 3 test cases, all pass (incl. the new statistics test).
- `unittests *IVF*` — 2096 assertions in 12 test cases, all pass.
- `make fmt` (clang-format 15) clean.

## Related Issues

- Closes: #2825

### CODE REVIEW

**Review by LHT129 — COMMENTED** [review]
Thanks for this PR. The change to move `GetStatistics` out of the `dist_cmp_reduce_mutex` critical section is well-motivated and correctly narrows the lock scope for the build path (where `ctx` is null, the entire block is now skipped) and the search path (where only the two integer accumulations remain serialized).

The test covers both pooled and non-pooled thread configurations and validates bucket-assignment parity as well as non-zero routing statistics — good coverage.

Two inline comments from the automated review are worth addressing:

1. **C++17 compatibility**: `QueryContext ctx{.stats = &stats}` uses a C++20 designated initializer. The project targets C++17 (per `cmake/VSAGCompilerConfig.cmake`), so this should be rewritten as `QueryContext ctx; ctx.stats = &stats;` or similar.

2. **`AddDistance` inside the mutex**: `SearchStatistics::AddDistance` is already lock-free (atomics + CAS). It could be called after releasing the mutex to further reduce contention, keeping only `dist_cmp += local_dist_cmp` inside the critical section.

File: src/algorithm/ivf/ivf_nearest_partition_test.cpp [hunk 1]
```diff
@@ -96,3 +96,41 @@ TEST_CASE("IVF Nearest Partition Serialize Test", "[ut][IVFNearestPartition]") {
...
+        dataset->Float32Vectors(vec.data())->Dim(dim)->NumElements(data_count)->Owner(false);
+
+        partition->Train(dataset);
+        auto class_result = partition->ClassifyDatas(vec.data(), data_count, 1, nullptr);
+
+        // The search path accumulates routing statistics through a non-null QueryContext. The
+        // statistics parsing must produce the same bucket assignment and non-zero routing stats
+        // (the parse now runs outside the reduce lock).
+        SearchStatistics stats;
+        QueryContext ctx{.stats = &stats};
```
[the remark this answers is not in the document]
Comment (LHT129) [reply src/algorithm/ivf/ivf_nearest_partition_test.cpp]: Fixed in 10523778: replaced C++20 designated initializer with C++17-compatible construct-and-assign pattern.

File: src/algorithm/ivf/ivf_nearest_partition.cpp [hunk 2]
```diff
@@ -114,15 +114,24 @@ IVFNearestPartition::ClassifyDatas(const void* datas,
...
+            if (route_stats.size() > 1) {
+                eval_count = std::strtoull(route_stats[1].c_str(), nullptr, 10);
+            }
+            {
+                std::scoped_lock lock(dist_cmp_reduce_mutex);
+                dist_cmp += local_dist_cmp;
+                ctx->stats->AddDistance(SearchStatistics::DistancePhase::ROUTING,
+                                        DistanceEvaluationBackend::FP32,
+                                        eval_count);
+            }
```
[the remark this answers is not in the document]
Comment (LHT129) [reply src/algorithm/ivf/ivf_nearest_partition.cpp:131]: Fixed in 10523778: moved AddDistance out of the scoped_lock; only the local dist_cmp accumulation remains inside the lock.

**Review by wxyucs — APPROVED** [review]
lgtm
